Подписка на других пользователей

SerUKalinin opened the pull request:
Подписка на других пользователей

SerUKalinin committed:
Cross Rewiew

InfernalBeard commented on src/main/java/school/faang/user_service/dto/UserFilterDto.java:
```diff
@@ -0,0 +1,21 @@
+    private int experienceMax;
+    private int page;
+    private int pageSize;
+}
```
РЕКОМЕНДАЦИЯ: использовать javax.validation для проверки значений полей.
ПРЕДЛОЖЕНИЕ: Добавить аннотации @Min, @Max, @NotNull и т.п. для валидации полей.

InfernalBeard commented on src/main/java/school/faang/user_service/dto/UserFilterDto.java:
```diff
@@ -0,0 +1,21 @@
+
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
```
РЕОКМЕНДАЦИЯ: Добавить аннотацию @Builder для удобного создания объектов. Это упростит код тестов и вызовов.

InfernalBeard commented on src/main/java/school/faang/user_service/dto/UserFilterDto.java:
```diff
@@ -0,0 +1,21 @@
+    private String phonePattern;
+    private String skillPattern;
+    private int experienceMin;
+    private int experienceMax;
```
РЕКОМЕНДАЦИЯ: Поля experienceMin и experienceMax также стоит валидировать, чтобы избежать ситуаций, когда experienceMin больше experienceMax.

InfernalBeard commented on src/main/java/school/faang/user_service/dto/UserDto.java:
```diff
@@ -0,0 +1,12 @@
+    private Long id;
+    private String username;
+    private String email;
+}
```
ПРЕДЛОЖЕНИЕ: Возможно добавить @AllArgsConstructor, чтобы сделать объект неизменяемым.

InfernalBeard commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,77 @@
+
+    @PostMapping("/follow")
+    public void followUser(@RequestParam long followerId,@RequestParam long followeeId) {
+        log.info("Attempt to follow user. FollowerId={}, followeeId={}", followerId, followeeId);
```
РЕКОМЕНДАЦИЯ: Проверка на существование подписки - добавить проверку, чтобы избежать дубликатов.
ПРЕДЛОЖЕНИЕ: Проверить на существование перед созданием новой подписки.

InfernalBeard commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,77 @@
+            log.info("Attempt self-subscription. FollowerId={}", followerId);
+            throw new IllegalArgumentException("You can't subscribe to yourself");
+        }
+        subscriptionService.followUser(followerId, followeeId);
```
РЕКОМЕНДАЦИЯ : Добавить обработку ошибок, чтобы система корректно реагировала на проблемы с базой данных.

InfernalBeard commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,77 @@
+       if(followerId == followeeId){
+           log.warn("Attempted self-unsubscription. FollowerId: {}", followerId);
+           throw new IllegalArgumentException("You can't unsubscribe from yourself");
+       }
```
РЕКОМЕНДАЦИЯ: Обернуть в try-catch и добавить логирование ошибок для надёжности.

InfernalBeard commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @GetMapping("/followers")
+    public List<UserDto> getFollowers(@RequestParam long followeeId, @RequestParam UserFilterDto filter) {
```
РЕКОМЕНДАЦИЯ: Добавить проверку валидности объекта filter перед вызовом сервиса.
ПРЕДЛОЖЕНИЕ: Валидация на уровне DTO и проверка на null.

InfernalBeard commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @GetMapping("/followees")
+    public List<UserDto> getFollowing(@RequestParam long followerId, @RequestParam UserFilterDto filter) {
```
РЕКОМЕНДАЦИЯ: Добавить валидацию объекта filter для корректной работы сервиса.
ПРЕДЛОЖЕНИЕ: Проверить filter и убедиться, что его поля корректны.

InfernalBeard commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public void followUser(long followerId, long followeeId) {
+        log.info("Attempting to follow: followerId={}, followeeId={}", followerId, followeeId);
```
РЕКОМЕНДАЦИЯ: Проверка на существование подписки - добавить проверку, чтобы избежать дубликатов.
ПРЕДЛОЖЕНИЕ: Проверить на существование перед созданием новой подписки.

InfernalBeard commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public void unfollowUser(long followerId, long followeeId) {
+        log.info("Attempting to unfollow: followerId={}, followeeId={}", followerId, followeeId);
```
РЕКОМЕНДАЦИЯ: Проверка на существование подписки перед удалением - исключает ненужные операции.
ПРЕДЛОЖЕНИЕ: Проверить, что подписка существует, прежде чем пытаться её удалить.

InfernalBeard commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,80 @@
+        if (!subscriptionRepository.existsByFollowerIdAndFolloweeId(followerId, followeeId)) {
+            log.warn("Subscription does not exist. FollowerId: {}, FolloweeId: {}", followerId, followeeId);
+            throw new IllegalArgumentException("Subscription does not exist");
+        }
```
РЕКОМЕНДАЦИЯ: Обернуть в try-catch и добавить логирование ошибок для надёжности.

InfernalBeard commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public List<UserDto> getFollowers(long followeeId, UserFilterDto filter) {
+        log.info("Getting followers for followeeId={}, filter={}", followeeId, filter);
```
РЕКОМЕНДАЦИЯ: Проверить валидность полей объекта filter перед использованием.
ПРЕДЛОЖЕНИЕ: Убедиться, что filter содержит корректные данные, чтобы избежать ошибок.

InfernalBeard commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,80 @@
+        log.info("FollowerId={} has followers", followeeId, count);
+        return count;
+    }
+
```
РЕКОМЕНДАЦИЯ: Функция фильтрации должна учитывать все возможные параметры из DTO.
ПРЕДЛОЖЕНИЕ: Добавить больше условий фильтрации.

InfernalBeard commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,80 @@
+
+    private List<UserDto> filterUsers(List<UserDto> users, UserFilterDto filter) {
+        log.info("Applying filters: {}", filter);
+        List<UserDto> filteredUsers = users.stream()
```
ПРАВИТЬ: Добавить дополнительные условия фильтрации по полям из UserFilterDto.

InfernalBeard reviewed (commented):
Именование методов и переменных:
Большинство имен методов и переменных имеют хорошую читаемость, но есть небольшие расхождения (например, getFolloweeCount в одном месте и getFollowingCount в другом для похожих целей). Стоит придерживаться более единообразного именования.
В контроллере используются термины followerId и followeeId. Это корректно, но для ясности можно рассмотреть использование слов follower и followed или что-то подобное.
Валидация:
Логика проверки подписки на самого себя и уже существующих подписок выглядит правильно, но лучше вынести её в отдельные методы, чтобы улучшить читаемость и повторное использование кода.
Рекомендуется использовать javax.validation для валидации входных данных в контроллере, чтобы избежать лишнего кода и упростить тестирование.
В методах фильтрации рекомендуется добавлять больше информации о применяемых фильтрах для лучшей отладки.
DTO и фильтрация
Реализация метода filterUsers в SubscriptionService слишком простая, она не учитывает все поля UserFilterDto. Следует доработать её, добавив фильтрацию по другим параметрам (например, по городу, навыкам и т.д.).

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,80 @@
+            .filter(user -> filter.getNamePattern() == null || user.getUsername().contains(filter.getNamePattern()))
+            .collect(Collectors.toList());
+        log.info("Filtered users count after applying filters: {}", filteredUsers.size());
+        return filteredUsers;
```
вместо создания новой структуры данных как лист, я бы сделал return результата стрима

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,80 @@
+
+    public List<UserDto> getFollowing(long followeeId, UserFilterDto filter) {
+        log.info("Fetching following list for followeeId={}, filter={}", followeeId, filter);
+        List<UserDto> following = (List<UserDto>) subscriptionRepository.findByFolloweeId(followeeId);
```
вмеcто явного приведения типов как `(List<UserDto>)`, я бы использовал конструкцию `subscriptionRepository.findByFolloweeId(followeeId).collect(Collectors.toList());` которая предназначена для таких нужд. То есть если ты используешь конструкцию `(List<UserDto>)` то это дает меньше гибкости чем метод `Collectors.toList()`

DaurenBerdibekov commented on src/main/java/school/faang/user_service/dto/UserDto.java:
```diff
@@ -0,0 +1,12 @@
+
+@Data
+@NoArgsConstructor
+public class UserDto {
```
я бы предложил написать DTO заглавными буквами, так как это аббревиатура.

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,80 @@
+        this.subscriptionRepository = subscriptionRepository;
+    }
+
+    public void followUser(long followerId, long followeeId) {
```
я бы предложил повесить на этот метод аннотацию @Transactional, чтобы спринг контроллировал целостность данных

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,80 @@
+        log.info("Successfully added: followerId={}, followeeId={}", followerId, followeeId);
+    }
+
+    public void unfollowUser(long followerId, long followeeId) {
```
и на этот тоже @Transactional

SerUKalinin committed:
Cross Rewiew

SerUKalinin committed:
Cross Rewiew

SerUKalinin committed:
Cross Rewiew

SerUKalinin committed:
Cross Rewiew

SerUKalinin committed:
Cross Rewiew

DaurenBerdibekov commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,107 @@
+
+    public SubscriptionController(SubscriptionService subscriptionService) {
+        this.subscriptionService = subscriptionService;
+    }
```
вместо такого конструктора предложил бы использовать аннотацию `@RequiredArgsConstructor`

DaurenBerdibekov commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    @PostMapping("/follow")
+    public ResponseEntity<String> followUser(@RequestParam Long followerId, @RequestParam Long followeeId) {
```
вместо `ResponseEntity` предложил бы использовать `@ResponseStatus` аннотацию. На счет всех методов в контроллере, я бы предложил вынести все валидации на сервисный уровень, так как из последнего митапа выяснилось, что контроллеры должны в идеале ТОЛЬКО принимать данные и отправлять их в сервис и получать отдуа ответ и отправлять клиенту.

DaurenBerdibekov commented on src/main/java/school/faang/user_service/dto/UserDTO.java:
```diff
@@ -0,0 +1,12 @@
+    private Long id;
+    private String username;
+    private String email;
+}
```
тут я бы тоже провалидировал поля

DaurenBerdibekov commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,107 @@
+    private boolean isValidFilter(UserFilterDTO filter) {
+        return filter.getExperienceMin() == null || filter.getExperienceMax() == null ||
+            filter.getExperienceMin() <= filter.getExperienceMax();
+    }
```
эти приватные методы тоже необходимо вынести в сервис

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,99 @@
+        this.subscriptionRepository = subscriptionRepository;
+    }
+
+    @Transactional
```
тут я бы убрал транзакцию, так как на последнем митапе выснилось что она не обязательна в методах со стандартными операциями с бд

zaqqaz226 replied:
Да, транзация далеко не всегда вешается над методами. которые ходят в базу ) Тут она не понадобится, тут у тебя простое действие и оборачивать в транзакцию нету смысла.

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,99 @@
+        log.info("Пользователь {} успешно подписался на пользователя {}.", followerId, followeeId);
+    }
+
+    @Transactional
```
и тут

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,99 @@
+        log.info("Запрос на получение подписчиков пользователя {}", userId);
+        if (filter == null) {
+            filter = new UserFilterDTO();
+        }
```
тут ты проверяешь `UserFilterDTO` является ли он `null` и это правильно. Но если он действительно `null`, то ты создаешь полностью новый объект этого дто. То есть, когда ты передаешь этот `filter` в метод `filterUsers()`, то он будет полностью пустым, потому что ты его создаешь с нуля. Поэтому я бы предложил вместо создания `new UserFilterDTO()`, выбрасывать исключение, что объект является `null`

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,99 @@
+        return followers;
+    }
+
+    public boolean subscriptionExists(Long followerId, Long followeeId) {
```
тут тоже можно было бы проверочку на входные параметры написать. И в других методах тоже

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,99 @@
+            .filter(user -> {
+                boolean matches = true;
+                if (filter.getNamePattern() != null) {
+                    matches = matches && new NameFilter(filter.getNamePattern()).filter(user);
```
мне кажется тут немного сложно читать, что тут происходит. Мне кажется было чуть понятнее, если бы все условия проверяли уже в if(...) а потом на основе этого переменая matches получает значение. 
```
if (filter.getNamePattern() != null && !new NameFilter(filter.getNamePattern()).filter(user)) {
                matches = false;
            }
```
. Было бы так дейсвительно лучше - я не знаю

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,99 @@
+                    log.info("Фильтр по имени с паттерном: {}", filter.getNamePattern());
+                }
+                if (filter.getExperienceMin() != null || filter.getExperienceMax() != null) {
+                    matches = matches && new ExperienceFilter(filter.getExperienceMin(), filter.getExperienceMax()).filter(user);
```
и тут тоже

DaurenBerdibekov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,99 @@
+            .map(user -> new UserDTO(user.getId(), user.getUsername(), user.getEmail()))
+            .peek(filteredUser -> log.info("Добавлен пользователь: {}", filteredUser.getUsername()))
+            .collect(Collectors.toList());
+    }
```
в этом методе ты хорошо используешь `stream` и сразу возращаешь результат, без создания дополнительных структур данных. Но в других методах, которые выше, перед тем как вернуть результат ты создаешь доп переменные. Например в методах `getFollowers` и `subscriptionExists` можно также сразу возвращать результат и не создавать доп переменные.

SerUKalinin committed:
Cross Rewiew

SerUKalinin committed:
Cross Rewiew

SerUKalinin committed:
Cross Rewiew

SerUKalinin committed:
Cross Rewiew

zaqqaz226 commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @DeleteMapping("/unfollow")
+    @ResponseStatus(HttpStatus.OK)
```
У тебя и так, если не будет выкинуто исключение, полетит 200. Считаю избыточным возвращаемый статус, если он просто 200.

zaqqaz226 commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,53 @@
+    @ResponseStatus(HttpStatus.OK)
+    public List<UserDTO> getFollowers(
+        @RequestParam Long userId,
+        @RequestBody(required = false) UserFilterDTO filter) {
```
Это у нас не обязательный параметр в методе?

zaqqaz226 commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,53 @@
+    @ResponseStatus(HttpStatus.OK)
+    public List<UserDTO> getFollowing(
+        @RequestParam Long followeeId,
+        @RequestBody(required = false) UserFilterDTO filter) {
```
Тот же вопрос что и выше)

zaqqaz226 commented on src/main/java/school/faang/user_service/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,37 @@
+
+@Slf4j
+@ControllerAdvice
+public class GlobalExceptionHandler {
```
Отлично, что сделал обработчик эксепшенов + логинование!

zaqqaz226 commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,108 @@
+            log.info("Пользователь {} успешно отписался от пользователя {}.", followerId, followeeId);
+        } catch (Exception ex) {
+            log.error("Произошла ошибка при отписке пользователя: followerId={}, followeeId={}", followerId, followeeId, ex);
+            throw new RuntimeException("Не удалось отписаться от пользователя.", ex);
```
Ты же уже создавал исключения свои, может под это тоже нужно свои исключение, а не выкидывать что-то общее?)

zaqqaz226 commented:
Отлично, исправляй все по комментам и жду тестов )

SerUKalinin committed:
Rewiew
- убрать @Transactional
- создать SubscriptionNotFoundException
- создать SubscriptionAlreadyExistsException
- изменить followUser
- изменить unfollowUser
-

SerUKalinin committed:
Rewiew
- убрать @Transactional
- создать SubscriptionNotFoundException
- создать SubscriptionAlreadyExistsException
- изменить followUser
- изменить unfollowUser
-

SerUKalinin committed:
Rewiew
- убрать @Transactional
- создать SubscriptionNotFoundException
- создать SubscriptionAlreadyExistsException
- изменить followUser
- изменить unfollowUser
-

SerUKalinin committed:
Rewiew
- убрать @Transactional
- создать SubscriptionNotFoundException
- создать SubscriptionAlreadyExistsException
- изменить followUser
- изменить unfollowUser
-

SerUKalinin committed:
Merge branch 'Titan-master-stream7' into Titan-master-stream7-BJS2-40509

SerUKalinin committed:
Последние изменения перед тестами

@ResponseStatus(HttpStatus.BAD_REQUEST)
класса: SubscriptionAlreadyExistsException
        SubscriptionNotFoundException

SerUKalinin committed:
Merge remote-tracking branch 'origin/Titan-master-stream7-BJS2-40509' into Titan-master-stream7-BJS2-40509

SerUKalinin committed:
Последние изменения перед тестами

в SubscriptionController  убрать @ResponseStatus(HttpStatus.OK) он является избыточным. По умолчанию, если метод контроллера выполняется без ошибок, Spring автоматически возвращает статус 200.

SerUKalinin committed:
Последние изменения перед тестами

добавил:
private String contactPattern;
    private String countryPattern;
    private String cityPattern;
    private String phonePattern;
    private String skillPattern;

    изменил метод filterUsers в SubscriptionService

    поменял в методе getFollowing
    followeeId на UserId

SerUKalinin committed:
Последние изменения перед тестами

добавил:
private String contactPattern;
    private String countryPattern;
    private String cityPattern;
    private String phonePattern;
    private String skillPattern;

    изменил метод filterUsers в SubscriptionService

    поменял в методе getFollowing
    followeeId на UserId

SerUKalinin committed:
Сделал тесты для SubscriptionService и SubscriptionController

SerUKalinin committed:
Сделал тесты для SubscriptionService и SubscriptionController

zaqqaz226 commented on src/main/java/school/faang/user_service/controllers/subscribe/SubscriptionController.java:
```diff
@@ -0,0 +1,50 @@
+    @GetMapping("/followers")
+    public List<UserDTO> getFollowers(
+        @RequestParam Long userId,
+        @RequestBody(required = false) UserFilterDTO filter) {
```
required = false, говорит о том, что этого параметра может не быть в запросе. Это точно не обязательный параметр?

zaqqaz226 commented on src/main/java/school/faang/user_service/controllers/subscribe/SubscriptionController.java:
```diff
@@ -0,0 +1,50 @@
+    @GetMapping("/following")
+    public List<UserDTO> getFollowing(
+        @RequestParam Long userId,
+        @RequestBody(required = false) UserFilterDTO filter) {
```
required = false, говорит о том, что этого параметра может не быть в запросе. Это точно не обязательный параметр?

zaqqaz226 commented on src/main/java/school/faang/user_service/services/subscribe/SubscriptionService.java:
```diff
@@ -0,0 +1,124 @@
+            log.info("Пользователь {} успешно отписался от пользователя {}.", followerId, followeeId);
+        } catch (Exception ex) {
+            log.error("Произошла ошибка при отписке пользователя: followerId={}, followeeId={}", followerId, followeeId, ex);
+            throw new RuntimeException("Не удалось отписаться от пользователя.", ex);
```
Почему везде твои исключения или более точные исключения, а тут общее оставил?

zaqqaz226 commented:
Тесты на контроллер, по идее, надо делать иначе. Могу сча это пропустить(так как это еще не изученный материал), но на следующей задаче, надо будет написать полноценные тесты через MockMvc. Решай когда займешься ими)

SerUKalinin committed:
Переделал тесты
Создал отдельно UnfollowException
Добавил @NoArgsConstructor в UserDTO
Убрал (required = false)

SerUKalinin committed:
в ParticipantRegistrationException и UnfollowException
добавил
@ResponseStatus(HttpStatus.BAD_REQUEST)

zaqqaz226 commented on src/main/java/school/faang/user_service/exceptions/subscribe/UnfollowException.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.BAD_REQUEST)
```
У тебя уже GlobalExceptionHandler и он отработает ответы, так что тут это лишнее. Как и в остальных исключениях

zaqqaz226 commented on src/test/java/school/faang/user_service/subscribe/controllers/SubscriptionControllerTest.java:
```diff
@@ -0,0 +1,98 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+@WebMvcTest(SubscriptionController.class)
+class SubscriptionControllerTest {
```
А они проходят?) Исходя из аннотаций не должны )

zaqqaz226 commented on src/test/java/school/faang/user_service/subscribe/controllers/SubscriptionControllerTest.java:
```diff
@@ -0,0 +1,98 @@
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+@WebMvcTest(SubscriptionController.class)
```
Тут как раз надо поставить @AutoConfigureMockMvc (или как то так)

zaqqaz226 commented on src/test/java/school/faang/user_service/subscribe/services/SubscriptionServiceTest.java:
```diff
@@ -0,0 +1,198 @@
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
+
+@AutoConfigureMockMvc
```
Тут должна быть иная аннотация. Эта отвечает за конфигурацию MockMvc, чтобы тестировать какие то http вызорвы. Например контроллер.

SerUKalinin committed:
Update GlobalExceptionHandler.java

SerUKalinin committed:
в ParticipantRegistrationException и UnfollowException
добавил
@ResponseStatus(HttpStatus.BAD_REQUEST)

SerUKalinin committed:
Merge remote-tracking branch 'origin/Titan-master-stream7-BJS2-40509' into Titan-master-stream7-BJS2-40509

SerUKalinin committed:
Добавил свой GlobalExceptionHandler в GlobalExceptionHandler даурена
